crates/fuel-core/src/graphql_api/validation_extension.rs

```diff
@@ -0,0 +1,97 @@
+use async_graphql::{
+    extensions::{
+        Extension,
+        ExtensionContext,
+        ExtensionFactory,
+        NextResolve,
+        ResolveInfo,
+    },
+    Pos,
+    ServerError,
+    ServerResult,
+    Value,
+};
+use std::{
+    collections::{
+        hash_map::Entry,
+        HashMap,
+    },
+    sync::Arc,
+};
+
+/// The extension validates the queries.
+pub(crate) struct ValidationExtension {
+    recursion_limit: usize,
+}
+
+impl ValidationExtension {
+    pub fn new(recursion_limit: usize) -> Self {
+        Self { recursion_limit }
+    }
+}
+
+impl ExtensionFactory for ValidationExtension {
+    fn create(&self) -> Arc<dyn Extension> {
+        Arc::new(ValidationInner::new(self.recursion_limit))
+    }
+}
+
+struct ValidationInner {
+    recursion_limit: usize,
+    visited: parking_lot::Mutex<HashMap<String, usize>>,
+}
+
+impl ValidationInner {
+    fn new(recursion_limit: usize) -> Self {
+        Self {
+            recursion_limit,
+            visited: parking_lot::Mutex::new(HashMap::new()),
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Extension for ValidationInner {
+    async fn resolve(
+        &self,
```

Comment:
Houston, we have a problem. This extension is shared between all concurrent query tasks. So if we have multiple requests for the same field, we may accidentally reject one query as recursive.

I tested it with this scenario:

```rust
#[tokio::test]
async fn submitting_multiple_queries_should_not_fail_but_it_does() {
    let query = r#"
    query {
      blocks(last: 1) {
        edges {
          cursor
          node {id}
        }
      }
    }"#;

    let node = FuelService::new_node(Config::local_node()).await.unwrap();
    let url = format!("http://{}/v1/graphql", node.bound_address);
    let client = FuelClient::new(url.clone()).unwrap();

    client
        .submit_and_await_commit(&Transaction::default_test_tx())
        .await
        .unwrap();

    for _ in 0..2 {
        let url = url.clone();
        let handle = tokio::spawn(async move {
            let result = send_graph_ql_query(&url, query).await;
            assert!(!result.contains("Recursion detected"), "{:?}", result);
        });
    }
}
```

which fails when inserting a ```tokio::time::sleep(std::time::Duration::from_millis(100)).await;``` before the `next.run` function.

Unfortunately I don't see a simple fix for this. We could perhaps do something with `tokio::task::id` ([link](https://docs.rs/tokio/latest/tokio/task/fn.id.html)) but that is an unstable feature.

Alternatively, this seems pretty unlikely (I could not trigger the error without inserting the artificial `sleep` call), so we could accept the risk for now and create a follow-up issue for the proper fix.

Reply:
It fails indeed, but not for the reason that you've mentioned=D

Each query has its own copy of the extension, so parallel query will have its own `visited: parking_lot::Mutex<HashMap<String, usize>>`.

But within one query, resolvers for other fields start to run when the resolver future is paused. And it is why test fails=)

```rust
#[tokio::test]
async fn submitting_multiple_queries_should_not_fail_but_it_does() {
    let query = r#"
    query {
      blocks(last: 1) {
        edges {
          cursor
          node {id}
        }
      }
    }"#;

    let node = FuelService::new_node(Config::local_node()).await.unwrap();
    let url = format!("http://{}/v1/graphql", node.bound_address);
    let client = FuelClient::new(url.clone()).unwrap();

    client
        .submit_and_await_commit(&Transaction::default_test_tx())
        .await
        .unwrap();
}
```

Reply:
Ah yeah you're right. There's no Arc, so there is no way the Mutex is shared between tasks. But still it's concurrently accessed like you say.

Reply:
Fixed with one time analyze during parsing of the query